### ISSUE TITLE
RegEx leaks memory when search() is called.

### ISSUE THREAD

**Issue by Atlinx** [godotengine/godot#50811]
### Godot version

v3.3.stable.mono

### System information

Windows 10

### Issue description

The `RegEx` class leaks memory whenever `RegEx.search()` is called.

### Steps to reproduce

1. Create a `RegEx` object and use `compile()` to configure the object with a regular expression pattern. 
2. Then in a `_process()` loop, call search() with some string.
3. Open up the debugger and monitor the static memory usage.
4. The static memory usage rises over time.

Code used in reproduction project:
![regex_memory_leak_script](https://user-images.githubusercontent.com/25368491/126875532-0de9fff8-f87c-4827-827f-32f393a467fd.png)

Monitor indicating memory leak:
![regex_memory_leak](https://user-images.githubusercontent.com/25368491/126875533-57b6ef78-4a69-4b8d-9796-9d6165540b8f.png)

Even if the `RegEx` object goes out of scope, the `search()` function will still have leaked memory:
![regex_memory_leak_alt_script](https://user-images.githubusercontent.com/25368491/126875647-26051648-1859-4e8c-919f-26c7f817b738.png)

Calling `RegEx.clear()` does not solve the memory leak either.

### Minimal reproduction project

[Regex Memory Leaking.zip](https://github.com/godotengine/godot/files/6872775/Regex.Memory.Leaking.zip)

**Comment by qarmin** [comment]
I cannot reproduce this leak on ubuntu.
RegEx is a Reference which is automatically cleared when it is not used.
Are you sure that this code produce leak?(I tested it in both mono and gdscript version)

**Comment by Atlinx** [comment]
> I cannot reproduce this leak on ubuntu.
> RegEx is a Reference which is automatically cleared when it is not used.
> Are you sure that this code produce leak?(I tested it in both mono and gdscript version)

Maybe this is a platform-dependent bug? I've run the test project on a fresh installation of Godot and the issue still persists.

Monitor of newly downloaded Godot editor running in self contained mode:
![fres_build](https://user-images.githubusercontent.com/25368491/126881758-5e551fd3-e5d6-4040-aa7b-40717108f446.png)

**Comment by Xrayez** [comment]
Maybe that's monitor's issue, something like `memnew`/`memdelete` mismatch?

For reference, Godot keeps track of memory allocations via custom `memnew`/`memdelete` macros. If thirdparty code (pcre2) frees memory via `free()` function and Godot allocates with `memnew()` and not `malloc()`, it may appear that memory keeps growing. But that's just an assumption.

**Comment by qarmin** [comment]
I tested this on Wine and leak is visible here

**Comment by GlyphTheWolf** [comment]
Hello all,

I tested this RegEx memory leak, because I used Godot for app which heavily use RegEx searching and I think fix for this memory leak is pretty easy. In regex.cpp match context is not freed before conditional exit (but only for case when `sizeof(CharType) == 2`):


```
(...)
if (sizeof(CharType) == 2) {
		pcre2_code_16 *c = (pcre2_code_16 *)code;
		pcre2_general_context_16 *gctx = (pcre2_general_context_16 *)general_ctx;
		pcre2_match_context_16 *mctx = pcre2_match_context_create_16(gctx);
		PCRE2_SPTR16 s = (PCRE2_SPTR16)p_subject.c_str();

		pcre2_match_data_16 *match = pcre2_match_data_create_from_pattern_16(c, gctx);

		int res = pcre2_match_16(c, s, length, p_offset, 0, match, mctx);

		if (res < 0) {
			pcre2_match_data_free_16(match);
			pcre2_match_context_free_16(mctx);  //This is missing
			return nullptr;
		}

		uint32_t size = pcre2_get_ovector_count_16(match);
		PCRE2_SIZE *ovector = pcre2_get_ovector_pointer_16(match);

		result->data.resize(size);

		for (uint32_t i = 0; i < size; i++) {
			result->data.write[i].start = ovector[i * 2];
			result->data.write[i].end = ovector[i * 2 + 1];
		}

		pcre2_match_data_free_16(match);
		pcre2_match_context_free_16(mctx);

	} else {
(...)
```

I tested this myself on Windows 10 64 bit with VisualStudio 2019 and without above fix I was able to reproduce issue. VisualStudio showed that leaked memory is from `pcre2_match_context_create_16` and with above fix issue doesn't reproduce for me. Maybe this will be helpful for someone.

**Comment by akien-mga** [comment]
If you want you can make a Pull Request with your proposed fix. See [PR workflow](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html) for details if you're not used to using Git / doing PRs on GitHub.

**Comment by akien-mga** [comment]
Fixed by #56763 (and for `master` it was fixed a while ago when refactoring String).
